File: dataforge-fx/src/main/java/hep/dataforge/fx/ApplicationSurrogate.java
package hep.dataforge.fx;

import javafx.application.Application;
import javafx.stage.Stage;

import java.util.concurrent.CompletableFuture;

/**
 * A surrogate JavaFX application to be used in case there is no global application present.
 * Created by darksnake on 23-Jan-17.
 */
public class ApplicationSurrogate extends Application {
    /**
     * The primary stage
     */
    private static CompletableFuture<Stage> stageGenerator;

    public static Stage getStage() {
        if (!isStarted()) {
            throw new RuntimeException("FX application surrogate not initialized");
        }
        try {
            return stageGenerator.get();
        } catch (Exception e) {
            throw new RuntimeException("Failed to load FX surrogate primary stage", e);
        }
    }

    public static boolean isStarted() {
        return stageGenerator != null;
    }

    public static void start() {
        stageGenerator = new CompletableFuture<>();
        new Thread(()->launch(ApplicationSurrogate.class)).start();
    }

    @Override
    public void start(Stage stage) throws Exception {
        stageGenerator.complete(stage);
    }

    @Override
    public void stop() throws Exception {
        stageGenerator = null;
        super.stop();
    }


}

File: kodex/dataforge-gui/src/main/kotlin/hep/dataforge/fx/ApplicationSurrogate.kt
package hep.dataforge.fx

import javafx.application.Application
import javafx.stage.Stage
import tornadofx.*

import java.util.concurrent.CompletableFuture

/**
 * A surrogate JavaFX application to be used in case there is no global application present.
 * Created by darksnake on 23-Jan-17.
 */
class ApplicationSurrogate : App() {

    @Throws(Exception::class)
    override fun start(stage: Stage) {
        stageGenerator!!.complete(stage)
    }

    @Throws(Exception::class)
    override fun stop() {
        stageGenerator = null
        super.stop()
    }

    companion object {
        /**
         * The primary stage
         */
        private var stageGenerator: CompletableFuture<Stage>? = null

        val stage: Stage
            get() {
                if (!isStarted) {
                    throw RuntimeException("FX application surrogate not initialized")
                }
                try {
                    return stageGenerator!!.get()
                } catch (e: Exception) {
                    throw RuntimeException("Failed to load FX surrogate primary stage", e)
                }

            }

        private val isStarted: Boolean
            get() = stageGenerator != null

        fun start() {
            if (!isStarted) {
                stageGenerator = CompletableFuture()
                Thread { Application.launch(ApplicationSurrogate::class.java) }.start()
            }
        }
    }


}
